Flaky unit test

/kind bug


This unit test is flaky: 

```
Based on the files in the current directory odo detected
Language: 
Project type: 
The devfile "" from the registry "" will be downloaded.
--- FAIL: TestInteractiveBackend_PersonalizeName (4.13s)
    --- FAIL: TestInteractiveBackend_PersonalizeName/no_flag (4.12s)
        controller.go:137: missing call(s) to *asker.MockAsker.AskName(is anything) /go/odo_1/pkg/init/backend/interactive_test.go:223
        controller.go:137: aborting test due to missing call(s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x279b492]

goroutine 70 [running]:
testing.tRunner.func1.2({0x321e380, 0x4948500})
	/usr/local/go/src/testing/testing.go:1209 +0x36c
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x3b6
panic({0x321e380, 0x4948500})
	/usr/local/go/src/runtime/panic.go:1047 +0x266
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.GetFilePathsFromRoot.func1({0xc00084e270, 0x17}, {0x0, 0x0}, {0xc000675070, 0xc0005100d0})
	/go/odo_1/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/language_recognizer.go:125 +0x132
path/filepath.walk({0x3488d3e, 0x4}, {0x38eede8, 0xc0005100d0}, 0xc0006b5070)
	/usr/local/go/src/path/filepath/path.go:438 +0x2a3
path/filepath.Walk({0x3488d3e, 0x4}, 0xc000675070)
	/usr/local/go/src/path/filepath/path.go:505 +0x85
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.GetFilePathsFromRoot({0x3488d3e, 0x4})
	/go/odo_1/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/language_recognizer.go:116 +0xdf
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.Analyze({0x3488d3e, 0x4})
	/go/odo_1/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/language_recognizer.go:36 +0x158
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.detectComponent({0x3488d3e, 0x4}, {0xc0006b5918, 0x0, 0x0})
	/go/odo_1/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:237 +0x9a
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.getComponentsWithoutConfigFile({0xc00038ed50, 0x1, 0x0})
	/go/odo_1/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:66 +0xee
github.com/redhat-developer/alizer/go/pkg/apis/recognizer.DetectComponents({0x3488d3e, 0x4})
	/go/odo_1/vendor/github.com/redhat-developer/alizer/go/pkg/apis/recognizer/component_recognizer.go:51 +0x8b
github.com/redhat-developer/odo/pkg/alizer.DetectName({0x3488d3e, 0x4})
	/go/odo_1/pkg/alizer/alizer.go:92 +0x94
github.com/redhat-developer/odo/pkg/init/backend.(*InteractiveBackend).PersonalizeName(0xc0006b5d38, {{{0x0, 0x0}, {0x3488d3e, 0x11}, {0x0, 0x0}, {0x0, 0x0, 0x0}, ...}, ...}, ...)
	/go/odo_1/pkg/init/backend/interactive.go:126 +0x69
github.com/redhat-developer/odo/pkg/init/backend.TestInteractiveBackend_PersonalizeName.func4(0xc0006d7040)
	/go/odo_1/pkg/init/backend/interactive_test.go:256 +0x2ca
testing.tRunner(0xc0006d7040, 0xc000456b40)
	/usr/local/go/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x727
FAIL	github.com/redhat-developer/odo/pkg/init/backend	4.797s
```

--- a/pkg/alizer/alizer.go
+++ b/pkg/alizer/alizer.go
@@ -47,13 +47,6 @@ func (o *Alizer) DetectFramework(path string) (recognizer.DevFileType, api.Regis
 	return types[typ], components.Items[typ].Registry, nil
 }
 
-func GetDevfileLocationFromDetection(typ recognizer.DevFileType, registry api.Registry) *api.DevfileLocation {
-	return &api.DevfileLocation{
-		Devfile:         typ.Name,
-		DevfileRegistry: registry.Name,
-	}
-}
-
 // DetectName retrieves the name of the project (if available)
 // If source code is detected:
 // 1. Detect the name (pom.xml for java, package.json for nodejs, etc.)
@@ -69,7 +62,7 @@ func GetDevfileLocationFromDetection(typ recognizer.DevFileType, registry api.Re
 // components, err := recognizer.DetectComponents("./")
 
 // In order to detect the name, the name will first try to find out the name based on the program (pom.xml, etc.) but then if not, it will use the dir name.
-func DetectName(path string) (string, error) {
+func (o *Alizer) DetectName(path string) (string, error) {
 	if path == "" {
 		return "", fmt.Errorf("path is empty")
 	}
@@ -119,4 +112,11 @@ func DetectName(path string) (string, error) {
 	}
 
 	return name, nil
+}
+
+func GetDevfileLocationFromDetection(typ recognizer.DevFileType, registry api.Registry) *api.DevfileLocation {
+	return &api.DevfileLocation{
+		Devfile:         typ.Name,
+		DevfileRegistry: registry.Name,
+	}
 }

--- a/pkg/alizer/alizer_test.go
+++ b/pkg/alizer/alizer_test.go
@@ -180,7 +180,11 @@ func TestDetectName(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 
-			name, err := DetectName(tt.args.path)
+			ctrl := gomock.NewController(t)
+			registryClient := registry.NewMockClient(ctrl)
+			alizerClient := NewAlizerClient(registryClient)
+
+			name, err := alizerClient.DetectName(tt.args.path)
 
 			if !tt.wantErr == (err != nil) {
 				t.Errorf("unexpected error %v, wantErr %v", err, tt.wantErr)

--- a/pkg/alizer/interface.go
+++ b/pkg/alizer/interface.go
@@ -7,4 +7,5 @@ import (
 
 type Client interface {
 	DetectFramework(path string) (recognizer.DevFileType, api.Registry, error)
+	DetectName(path string) (string, error)
 }

--- a/pkg/alizer/mock.go
+++ b/pkg/alizer/mock.go
@@ -50,3 +50,18 @@ func (mr *MockClientMockRecorder) DetectFramework(path interface{}) *gomock.Call
 	mr.mock.ctrl.T.Helper()
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "DetectFramework", reflect.TypeOf((*MockClient)(nil).DetectFramework), path)
 }
+
+// DetectName mocks base method.
+func (m *MockClient) DetectName(path string) (string, error) {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "DetectName", path)
+	ret0, _ := ret[0].(string)
+	ret1, _ := ret[1].(error)
+	return ret0, ret1
+}
+
+// DetectName indicates an expected call of DetectName.
+func (mr *MockClientMockRecorder) DetectName(path interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "DetectName", reflect.TypeOf((*MockClient)(nil).DetectName), path)
+}

--- a/pkg/component/component.go
+++ b/pkg/component/component.go
@@ -57,7 +57,8 @@ func GatherName(contextDir string, devfileObj *parser.DevfileObj) (string, error
 			// Use Alizer if Devfile has no (optional) metadata.name field.
 			// We need to pass in the Devfile base directory (not the path to the devfile.yaml).
 			// Name returned by alizer.DetectName is expected to be already sanitized.
-			return alizer.DetectName(filepath.Dir(devfileObj.Ctx.GetAbsPath()))
+			alizerClient := alizer.Alizer{} // TODO(feloy) fix with DI
+			return alizerClient.DetectName(filepath.Dir(devfileObj.Ctx.GetAbsPath()))
 		}
 	} else {
 		// Fallback to the context dir name

--- a/pkg/init/backend/alizer.go
+++ b/pkg/init/backend/alizer.go
@@ -58,7 +58,7 @@ func (o *AlizerBackend) PersonalizeName(devfile parser.DevfileObj, flags map[str
 	if path == "" {
 		return "", fmt.Errorf("cannot determine the absolute path of the directory")
 	}
-	return alizer.DetectName(path)
+	return o.alizerClient.DetectName(path)
 }
 
 func (o *AlizerBackend) PersonalizeDevfileConfig(devfile parser.DevfileObj) (parser.DevfileObj, error) {

--- a/pkg/init/backend/interactive.go
+++ b/pkg/init/backend/interactive.go
@@ -29,14 +29,16 @@ const (
 type InteractiveBackend struct {
 	askerClient    asker.Asker
 	registryClient registry.Client
+	alizerClient   alizer.Client
 }
 
 var _ InitBackend = (*InteractiveBackend)(nil)
 
-func NewInteractiveBackend(askerClient asker.Asker, registryClient registry.Client) *InteractiveBackend {
+func NewInteractiveBackend(askerClient asker.Asker, registryClient registry.Client, alizerClient alizer.Client) *InteractiveBackend {
 	return &InteractiveBackend{
 		askerClient:    askerClient,
 		registryClient: registryClient,
+		alizerClient:   alizerClient,
 	}
 }
 
@@ -123,7 +125,7 @@ func (o *InteractiveBackend) PersonalizeName(devfile parser.DevfileObj, flags ma
 	baseDir := filepath.Dir(path)
 
 	// Detect the name
-	name, err := alizer.DetectName(baseDir)
+	name, err := o.alizerClient.DetectName(baseDir)
 	if err != nil {
 		return "", fmt.Errorf("detecting name using alizer: %w", err)
 	}

--- a/pkg/init/backend/interactive_test.go
+++ b/pkg/init/backend/interactive_test.go
@@ -6,6 +6,7 @@ import (
 
 	"github.com/golang/mock/gomock"
 
+	"github.com/redhat-developer/odo/pkg/alizer"
 	"github.com/redhat-developer/odo/pkg/api"
 	"github.com/redhat-developer/odo/pkg/init/asker"
 	"github.com/redhat-developer/odo/pkg/registry"
@@ -203,6 +204,7 @@ func TestInteractiveBackend_PersonalizeName(t *testing.T) {
 	type fields struct {
 		asker          func(ctrl *gomock.Controller) asker.Asker
 		registryClient registry.Client
+		alizer         func(ctrl *gomock.Controller) alizer.Client
 	}
 	type args struct {
 		devfile func(fs filesystem.Filesystem) parser.DevfileObj
@@ -221,6 +223,11 @@ func TestInteractiveBackend_PersonalizeName(t *testing.T) {
 				asker: func(ctrl *gomock.Controller) asker.Asker {
 					client := asker.NewMockAsker(ctrl)
 					client.EXPECT().AskName(gomock.Any()).Return("aname", nil)
+					return client
+				},
+				alizer: func(ctrl *gomock.Controller) alizer.Client {
+					client := alizer.NewMockClient(ctrl)
+					client.EXPECT().DetectName(gomock.Any()).Return("name1", nil)
 					return client
 				},
 			},
@@ -248,9 +255,14 @@ func TestInteractiveBackend_PersonalizeName(t *testing.T) {
 			if tt.fields.asker != nil {
 				askerClient = tt.fields.asker(ctrl)
 			}
+			var alizerClient alizer.Client
+			if tt.fields.alizer != nil {
+				alizerClient = tt.fields.alizer(ctrl)
+			}
 			o := &InteractiveBackend{
 				askerClient:    askerClient,
 				registryClient: tt.fields.registryClient,
+				alizerClient:   alizerClient,
 			}
 			fs := filesystem.NewFakeFs()
 			newName, err := o.PersonalizeName(tt.args.devfile(fs), tt.args.flags)

--- a/pkg/init/init.go
+++ b/pkg/init/init.go
@@ -44,7 +44,7 @@ func NewInitClient(fsys filesystem.Filesystem, preferenceClient preference.Clien
 	askerClient := asker.NewSurveyAsker()
 	return &InitClient{
 		flagsBackend:       backend.NewFlagsBackend(preferenceClient),
-		interactiveBackend: backend.NewInteractiveBackend(askerClient, registryClient),
+		interactiveBackend: backend.NewInteractiveBackend(askerClient, registryClient, alizerClient),
 		alizerBackend:      backend.NewAlizerBackend(askerClient, alizerClient),
 		fsys:               fsys,
 		preferenceClient:   preferenceClient,

